File: packages/app/src/emr/services/observability/phiScrubber.ts
// SPDX-FileCopyrightText: Copyright LiverRa contributors
// SPDX-License-Identifier: Apache-2.0

/**
 * PHI scrubber — browser port of `packages/ml-inference/src/observability/phi_scrubber.py` (T070).
 *
 * Wired into Sentry + PostHog `beforeSend` hooks. Fail-closed per spec
 * FR-029b: if any rule throws, we raise `ScrubberFailure` and the
 * calling hook MUST drop the event (return null) rather than send it.
 *
 * Plain-English analogy:
 *   A shredder at the outbox. Every event is passed through — names,
 *   patient IDs, emails are blacked out. If the shredder jams, the
 *   outbox throws the envelope away. Better to lose a crash report
 *   than leak PHI.
 *
 * TODO(sync): keep in sync with the Python scrubber. The name lists +
 * regexes below are deliberately identical to their Python cousins.
 * A future codegen step should derive both from a shared JSON source.
 */

export class ScrubberFailure extends Error {
  override readonly name = 'ScrubberFailure';
}

export const REDACTION = '[redacted]' as const;

/** Fields whose values pass through unchanged. Matches `SAFE_FIELD_NAMES` in Python. */
export const SAFE_FIELD_NAMES: ReadonlySet<string> = new Set([
  'analysis_id',
  'correlation_id',
  'error_code',
  'error_slug',
  'event_type',
  'leaf_hash',
  'level',
  'model_version',
  'pipeline_stage',
  'prev_leaf_hash',
  'request_id',
  'sequence_no',
  'series_instance_uid',
  'sop_instance_uid',
  'stage',
  'study_instance_uid',
  'tenant_id',
  'timestamp',
  'trace_id',
  'transaction_uid',
  'user_role',
]);

// Regexes — keep mirror-identical to the Python tree.
const DICOM_UID_RE = /\b(?:\d+\.){8,}\d+\b/g;
const EMAIL_RE = /\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}\b/gi;
const MRN_LABEL_RE =
  /\b(?:MRN|Patient\s*ID|PatientID|Patient\s*No\.?|Pat\.?\s*Nr\.?|Chart\s*No\.?|Medical\s*Record\s*Number)\s*[:#]?\s*[A-Z0-9\-/]{3,}/gi;
const BARE_MRN_RE = /\b\d{6,10}\b/g;

/**
 * PARTIAL LIST — production deployment should load full list from tenant config.
 * German top-50 given + family names (Destatis 2023, Deutsche Post Namensforschung).
 */
export const GERMAN_NAMES: readonly string[] = [
  // Given
  'Alexander', 'Andreas', 'Benjamin', 'Christian', 'Daniel',
  'David', 'Dennis', 'Dominik', 'Felix', 'Florian',
  'Hans', 'Jakob', 'Jan', 'Johannes', 'Jonas',
  'Julian', 'Kevin', 'Leon', 'Lukas', 'Markus',
  'Martin', 'Matthias', 'Maximilian', 'Michael', 'Niklas',
  'Paul', 'Peter', 'Philipp', 'Sebastian', 'Simon',
  'Stefan', 'Thomas', 'Tim', 'Tobias',
  'Anna', 'Elena', 'Emma', 'Hannah', 'Johanna',
  'Julia', 'Katharina', 'Laura', 'Lea', 'Lena',
  'Lisa', 'Maria', 'Marie', 'Nina', 'Petra',
  'Sabine', 'Sarah', 'Sophie', 'Stefanie',
  // Family
  'Müller', 'Mueller', 'Schmidt', 'Schneider', 'Fischer',
  'Weber', 'Meyer', 'Wagner', 'Becker', 'Schulz',
  'Hoffmann', 'Schäfer', 'Schaefer', 'Koch', 'Bauer',
  'Richter', 'Klein', 'Wolf', 'Schröder', 'Schroeder',
  'Neumann', 'Schwarz', 'Zimmermann', 'Braun', 'Krüger',
  'Krueger', 'Hofmann', 'Hartmann', 'Lange', 'Werner',
  'Schmitz', 'Krause', 'Meier', 'Lehmann', 'Schmid',
  'Schulze', 'Maier', 'Köhler', 'Koehler', 'Herrmann',
  'König', 'Koenig', 'Walter', 'Mayer', 'Huber',
  'Kaiser', 'Fuchs', 'Peters', 'Lang', 'Scholz',
  'Möller', 'Moeller', 'Weiß', 'Weiss', 'Jung',
  'Hahn', 'Schubert', 'Özdemir', 'Oezdemir',
];

/**
 * PARTIAL LIST — production deployment should load full list from tenant config.
 * Georgian given + family names in Latin transliteration AND native script.
 */
export const GEORGIAN_NAMES: readonly string[] = [
  // Latin
  'Giorgi', 'Davit', 'Irakli', 'Levan', 'Zviad',
  'Nika', 'Lasha', 'Vakhtang', 'Tornike', 'Mikheil',
  'Nodar', 'Shota', 'Otar', 'Zurab', 'Revaz',
  'Beka', 'Sandro', 'Luka',
  'Nino', 'Tamar', 'Mariam', 'Ana', 'Salome',
  'Lika', 'Elene', 'Natia', 'Ketevan', 'Tamta',
  'Mzia', 'Manana',
  'Gogichaishvili', 'Giorgadze', 'Svanadze', 'Japaridze',
  'Beridze', 'Kapanadze', 'Tsereteli', 'Kiknadze',
  'Chavchavadze', 'Saakashvili', 'Gelashvili', 'Lomidze',
  'Khutsishvili', 'Meladze', 'Tabatadze', 'Gurgenidze',
  'Kakabadze', 'Lortkipanidze', 'Dolidze', 'Nadiradze',
  'Chkheidze', 'Tevzadze', 'Khelaia', 'Kvaratskhelia',
  // Georgian script
  'გიორგი', 'დავით', 'ირაკლი', 'ლევან', 'ზვიად',
  'ნიკა', 'ლაშა', 'ვახტანგ', 'თორნიკე', 'მიხეილ',
  'ნინო', 'თამარ', 'მარიამ', 'ანა', 'სალომე',
  'ლიკა', 'ელენე', 'ნათია', 'ქეთევან',
  'გოგიჩაიშვილი', 'გიორგაძე', 'სვანაძე', 'ჯაფარიძე',
  'ბერიძე', 'კაპანაძე', 'წერეთელი', 'კიკნაძე',
  'ჭავჭავაძე', 'სააკაშვილი', 'გელაშვილი', 'ლომიძე',
];

// Pre-compute a longest-first name list so multi-token names (e.g.
// "Meyer-Schulze") are caught before single tokens that would shadow them.
const NAME_LIST: readonly string[] = [...new Set([...GERMAN_NAMES, ...GEORGIAN_NAMES])]
  .sort((a, b) => b.length - a.length);

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Scrub a string of PHI (names, MRNs, emails, DICOM UIDs).
 * Throws `ScrubberFailure` if any rule crashes — caller MUST drop the event.
 */
export function scrubString(input: string): string {
  try {
    if (!input) return input;

    let out = input;

    // 1. DICOM UIDs.
    out = out.replace(DICOM_UID_RE, REDACTION);

    // 2. MRN labels.
    out = out.replace(MRN_LABEL_RE, REDACTION);

    // 3. Emails.
    out = out.replace(EMAIL_RE, REDACTION);

    // 4. Bare MRN-shaped digit runs.
    out = out.replace(BARE_MRN_RE, REDACTION);

    // 5. Names — longest-first linear scan with word-boundary check.
    for (const name of NAME_LIST) {
      if (!out.includes(name)) continue;
      // Word boundaries: previous + next char must be non-letter.
      out = replaceWithWordBoundary(out, name, REDACTION);
    }

    return out;
  } catch (err) {
    throw new ScrubberFailure(
      `phi_scrubber.scrubString failed: ${(err as Error)?.message ?? String(err)}`,
    );
  }
}

/** Deep-clone + scrub an arbitrary object graph. Fail-closed. */
export function scrubObject<T>(input: T): T {
  try {
    return walk(input, undefined) as T;
  } catch (err) {
    if (err instanceof ScrubberFailure) throw err;
    throw new ScrubberFailure(
      `phi_scrubber.scrubObject failed: ${(err as Error)?.message ?? String(err)}`,
    );
  }
}

// ---------------------------------------------------------------------------
// Internals
// ---------------------------------------------------------------------------

function walk(node: unknown, parentKey: string | undefined): unknown {
  if (node === null || node === undefined) return node;
  if (Array.isArray(node)) return node.map((v) => walk(v, parentKey));
  if (typeof node === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(node as Record<string, unknown>)) {
      out[k] = walk(v, k);
    }
    return out;
  }
  if (typeof node === 'string') {
    if (parentKey !== undefined && SAFE_FIELD_NAMES.has(parentKey)) {
      return node;
    }
    return scrubString(node);
  }
  return node;
}

/**
 * Replace every occurrence of `needle` in `haystack` with `replacement`,
 * but only when bounded by non-letter characters (Unicode-aware).
 */
function replaceWithWordBoundary(
  haystack: string,
  needle: string,
  replacement: string,
): string {
  let out = '';
  let i = 0;
  while (i < haystack.length) {
    const j = haystack.indexOf(needle, i);
    if (j < 0) {
      out += haystack.slice(i);
      break;
    }
    const prevChar = j === 0 ? '' : haystack[j - 1];
    const nextChar = haystack[j + needle.length] ?? '';
    const prevOk = !prevChar || !isLetter(prevChar);
    const nextOk = !nextChar || !isLetter(nextChar);
    if (prevOk && nextOk) {
      out += haystack.slice(i, j) + replacement;
      i = j + needle.length;
    } else {
      out += haystack.slice(i, j + 1);
      i = j + 1;
    }
  }
  return out;
}

function isLetter(ch: string): boolean {
  // Unicode property escape — letter OR combining mark (Georgian + German-diacritics aware).
  return /\p{L}|\p{M}/u.test(ch);
}

export default { scrubString, scrubObject, ScrubberFailure };

File: packages/ml-inference/src/observability/phi_scrubber.py
# SPDX-FileCopyrightText: Copyright LiverRa contributors
# SPDX-License-Identifier: Apache-2.0
"""PHI scrubber (T069).

Rules-based scrubber for Sentry / PostHog / structured-log payloads
(LiverRa spec NFR-007). Fail-closed per FR-029b: if any exception is
raised during scrubbing the event MUST be dropped, never sent.

Plain-English analogy:
    Imagine a shredder that sits in front of the mailbox. Every letter
    going out is first fed through — names, patient IDs, email
    addresses are blacked out. If the shredder jams, the letter never
    leaves the building. Better to lose observability than leak PHI.

Rules implemented in v1:

1. DICOM UIDs           — dotted numeric ≥9 components, length 16–64
2. Emails               — RFC-5322-ish
3. MRN-like tokens      — configurable per tenant (e.g. "MRN: 1234567",
                          "Patient ID: X-5532/2025")
4. German given/family name list (top 50, partial — see constants)
5. Georgian given/family name list (Latin + native script)
6. Field-name allowlist — keys that are safe even if their values look
   like MRNs (``study_instance_uid``, ``analysis_id``, ``model_version``,
   ``error_slug``, ``stage``, ``sequence_no``, ``tenant_id``, …).
"""
from __future__ import annotations

import copy
import logging
import re
from dataclasses import dataclass, field
from typing import Any, Iterable

from prometheus_client import Counter

try:  # pragma: no cover — exercised at import time.
    import ahocorasick  # type: ignore[import-untyped]
except ImportError:  # pragma: no cover
    ahocorasick = None  # type: ignore[assignment]

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Metrics
# ---------------------------------------------------------------------------

phi_scrubber_failed_total = Counter(
    "phi_scrubber_failed_total",
    "Count of scrubber failures (fail-closed — event dropped).",
    labelnames=("reason",),
)


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

REDACTION: str = "[redacted]"

# Field names whose VALUES pass through unchanged even if they look
# MRN-shaped. Everything else is scrubbed.
SAFE_FIELD_NAMES: frozenset[str] = frozenset(
    {
        "analysis_id",
        "correlation_id",
        "error_code",
        "error_slug",
        "event_type",
        "leaf_hash",
        "level",
        "model_version",
        "pipeline_stage",
        "prev_leaf_hash",
        "request_id",
        "sequence_no",
        "series_instance_uid",
        "sop_instance_uid",
        "stage",
        "study_instance_uid",
        "tenant_id",
        "timestamp",
        "trace_id",
        "transaction_uid",
        "user_role",
    }
)

# DICOM UID — dotted numeric, at least 9 groups, 16–64 chars.
# NOTE: only scrub DICOM UIDs OUTSIDE of allowlisted fields; inside an
# allowlisted field (e.g. ``study_instance_uid``) they're considered safe.
_DICOM_UID_RE = re.compile(r"\b(?:\d+\.){8,}\d+\b")

_EMAIL_RE = re.compile(
    r"(?i)\b[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}\b"
)

# MRN-like labels. Case-insensitive, matches the LABEL + its payload.
_MRN_LABEL_RE = re.compile(
    r"(?i)\b(?:MRN|Patient\s*ID|PatientID|Patient\s*No\.?|Pat\.?\s*Nr\.?|"
    r"Chart\s*No\.?|Medical\s*Record\s*Number)\s*[:#]?\s*[A-Z0-9\-/]{3,}"
)

# Bare 6-10 digit numeric bursts that are likely MRNs when NOT in a
# safe field. Intentionally conservative — we do not touch sequence_no
# values because they are recognized via field name.
_BARE_MRN_RE = re.compile(r"\b\d{6,10}\b")

# --- German given + family names (PARTIAL LIST — production deployment
# should load the full curated dataset from tenant config). Source: public
# top-50 lists (Destatis 2023, Deutsche Post Namensforschung).
# PARTIAL LIST — production deployment should load full list from tenant config
GERMAN_NAMES: tuple[str, ...] = (
    # Given (male)
    "Alexander", "Andreas", "Benjamin", "Christian", "Daniel",
    "David", "Dennis", "Dominik", "Felix", "Florian",
    "Hans", "Jakob", "Jan", "Johannes", "Jonas",
    "Julian", "Kevin", "Leon", "Lukas", "Markus",
    "Martin", "Matthias", "Maximilian", "Michael", "Niklas",
    "Paul", "Peter", "Philipp", "Sebastian", "Simon",
    "Stefan", "Thomas", "Tim", "Tobias",
    # Given (female)
    "Anna", "Elena", "Emma", "Hannah", "Johanna",
    "Julia", "Katharina", "Laura", "Lea", "Lena",
    "Lisa", "Maria", "Marie", "Nina", "Petra",
    "Sabine", "Sarah", "Sophie", "Stefanie",
    # Family (top 50)
    "Müller", "Mueller", "Schmidt", "Schneider", "Fischer",
    "Weber", "Meyer", "Wagner", "Becker", "Schulz",
    "Hoffmann", "Schäfer", "Schaefer", "Koch", "Bauer",
    "Richter", "Klein", "Wolf", "Schröder", "Schroeder",
    "Neumann", "Schwarz", "Zimmermann", "Braun", "Krüger",
    "Krueger", "Hofmann", "Hartmann", "Lange", "Werner",
    "Schmitz", "Krause", "Meier", "Lehmann", "Schmid",
    "Schulze", "Maier", "Köhler", "Koehler", "Herrmann",
    "König", "Koenig", "Walter", "Mayer", "Huber",
    "Kaiser", "Fuchs", "Peters", "Lang", "Scholz",
    "Möller", "Moeller", "Weiß", "Weiss", "Jung",
    "Hahn", "Schubert", "Özdemir", "Oezdemir",
)

# --- Georgian names (PARTIAL LIST). Both Latin transliteration and
# native script. Source: Statistics Georgia + hospital registries.
# PARTIAL LIST — production deployment should load full list from tenant config
GEORGIAN_NAMES: tuple[str, ...] = (
    # Given (male) — Latin
    "Giorgi", "Davit", "Irakli", "Levan", "Zviad",
    "Nika", "Lasha", "Vakhtang", "Tornike", "Mikheil",
    "Nodar", "Shota", "Otar", "Zurab", "Revaz",
    "Beka", "Sandro", "Luka",
    # Given (female) — Latin
    "Nino", "Tamar", "Mariam", "Ana", "Salome",
    "Lika", "Elene", "Natia", "Ketevan", "Tamta",
    "Mzia", "Manana",
    # Family — Latin (typical -shvili / -dze / -adze endings)
    "Gogichaishvili", "Giorgadze", "Svanadze", "Japaridze",
    "Beridze", "Kapanadze", "Tsereteli", "Kiknadze",
    "Chavchavadze", "Saakashvili", "Gelashvili", "Lomidze",
    "Khutsishvili", "Meladze", "Tabatadze", "Gurgenidze",
    "Kakabadze", "Lortkipanidze", "Dolidze", "Nadiradze",
    "Chkheidze", "Tevzadze", "Khelaia", "Kvaratskhelia",
    # Native script — common given + family
    "გიორგი", "დავით", "ირაკლი", "ლევან", "ზვიად",
    "ნიკა", "ლაშა", "ვახტანგ", "თორნიკე", "მიხეილ",
    "ნინო", "თამარ", "მარიამ", "ანა", "სალომე",
    "ლიკა", "ელენე", "ნათია", "ქეთევან",
    "გოგიჩაიშვილი", "გიორგაძე", "სვანაძე", "ჯაფარიძე",
    "ბერიძე", "კაპანაძე", "წერეთელი", "კიკნაძე",
    "ჭავჭავაძე", "სააკაშვილი", "გელაშვილი", "ლომიძე",
)


class ScrubberFailure(RuntimeError):
    """Raised when scrubbing fails. Caller MUST drop the event."""


# ---------------------------------------------------------------------------
# Scrubber implementation
# ---------------------------------------------------------------------------


@dataclass
class PHIScrubber:
    """Rules-based, fail-closed PHI scrubber.

    ``PHIScrubber`` is intentionally cheap to construct — the heavy
    state (compiled Aho-Corasick automaton) is built lazily.
    """

    extra_name_list: tuple[str, ...] = field(default_factory=tuple)
    extra_mrn_patterns: tuple[str, ...] = field(default_factory=tuple)
    safe_field_names: frozenset[str] = field(default_factory=lambda: SAFE_FIELD_NAMES)

    _automaton: Any = field(init=False, repr=False, default=None)
    _extra_mrn_re: re.Pattern[str] | None = field(init=False, repr=False, default=None)

    def __post_init__(self) -> None:
        self._automaton = self._build_automaton(
            GERMAN_NAMES + GEORGIAN_NAMES + self.extra_name_list
        )
        if self.extra_mrn_patterns:
            joined = "|".join(f"(?:{p})" for p in self.extra_mrn_patterns)
            self._extra_mrn_re = re.compile(joined)

    # -- public API ------------------------------------------------------

    def scrub_dict(self, obj: Any) -> Any:
        """Deep-copy ``obj`` and redact PHI. Fail-closed.

        Walks all nested dicts / lists / tuples. Any exception is
        translated into :class:`ScrubberFailure` so the caller (Sentry
        ``before_send`` etc.) can drop the event.
        """
        try:
            cloned = copy.deepcopy(obj)
            return self._walk(cloned, parent_key=None)
        except ScrubberFailure:
            raise
        except Exception as exc:  # noqa: BLE001 — fail-closed is the contract.
            phi_scrubber_failed_total.labels(reason="scrub_dict").inc()
            logger.error("phi_scrubber.scrub_dict failed: %s", exc, exc_info=True)
            raise ScrubberFailure("scrub_dict failed") from exc

    def scrub_string(self, s: str) -> str:
        """Scrub a free-text string. Fail-closed."""
        try:
            return self._scrub_string_inner(s)
        except Exception as exc:  # noqa: BLE001
            phi_scrubber_failed_total.labels(reason="scrub_string").inc()
            logger.error(
                "phi_scrubber.scrub_string failed: %s", exc, exc_info=True
            )
            raise ScrubberFailure("scrub_string failed") from exc

    # -- internals -------------------------------------------------------

    def _walk(self, node: Any, *, parent_key: str | None) -> Any:
        if isinstance(node, dict):
            return {k: self._walk(v, parent_key=k) for k, v in node.items()}
        if isinstance(node, list):
            return [self._walk(v, parent_key=parent_key) for v in node]
        if isinstance(node, tuple):
            return tuple(self._walk(v, parent_key=parent_key) for v in node)
        if isinstance(node, str):
            if parent_key and parent_key in self.safe_field_names:
                # Field name is on the allowlist — leave value alone.
                return node
            return self._scrub_string_inner(node)
        # Non-string scalars (int/float/bool/None) → pass through.
        return node

    def _scrub_string_inner(self, s: str) -> str:
        if not s:
            return s

        # 1. DICOM UIDs.
        redacted = _DICOM_UID_RE.sub(REDACTION, s)

        # 2. MRN labels (redact the ENTIRE label+payload span).
        redacted = _MRN_LABEL_RE.sub(REDACTION, redacted)

        # 3. Tenant-specific MRN patterns (if any).
        if self._extra_mrn_re is not None:
            redacted = self._extra_mrn_re.sub(REDACTION, redacted)

        # 4. Emails.
        redacted = _EMAIL_RE.sub(REDACTION, redacted)

        # 5. Bare 6-10 digit numeric bursts (residual MRNs).
        redacted = _BARE_MRN_RE.sub(REDACTION, redacted)

        # 6. Names (Aho-Corasick). We replace token-by-token: the
        #    automaton gives us (end_index, value). We sort matches
        #    by descending end_index so earlier replacements don't
        #    shift indices for later ones.
        redacted = self._redact_names(redacted)

        return redacted

    def _redact_names(self, s: str) -> str:
        if self._automaton is None:
            return s

        # Collect non-overlapping matches with simple word-boundary check.
        matches: list[tuple[int, int]] = []  # (start, end)
        try:
            iter_ = self._automaton.iter(s)
        except Exception:  # noqa: BLE001
            # Automaton returned before finalization — treat as no matches.
            return s

        for end_index, value in iter_:
            start = end_index - len(value) + 1
            end = end_index + 1
            # Word-boundary: previous char not alpha, next char not alpha
            prev_ok = start == 0 or not s[start - 1].isalpha()
            next_ok = end == len(s) or not s[end].isalpha()
            if prev_ok and next_ok:
                matches.append((start, end))

        if not matches:
            return s

        # Merge overlaps + replace from right to left.
        matches.sort()
        merged: list[tuple[int, int]] = []
        for start, end in matches:
            if merged and start <= merged[-1][1]:
                merged[-1] = (merged[-1][0], max(merged[-1][1], end))
            else:
                merged.append((start, end))

        out = s
        for start, end in reversed(merged):
            out = out[:start] + REDACTION + out[end:]
        return out

    @staticmethod
    def _build_automaton(names: Iterable[str]) -> Any:
        if ahocorasick is None:
            # Fallback: simple set + linear scan in _redact_names via regex.
            # We still return an object with an ``iter`` method for uniform code.
            class _FallbackAutomaton:
                def __init__(self, words: tuple[str, ...]) -> None:
                    # Longest-first so multi-word names are caught before
                    # single tokens.
                    self._patterns = sorted(set(words), key=len, reverse=True)

                def iter(self, s: str) -> Iterable[tuple[int, str]]:
                    for pat in self._patterns:
                        start = 0
                        while True:
                            i = s.find(pat, start)
                            if i < 0:
                                break
                            yield (i + len(pat) - 1, pat)
                            start = i + len(pat)

            return _FallbackAutomaton(tuple(names))

        auto = ahocorasick.Automaton()
        for name in names:
            if name:
                auto.add_word(name, name)
        auto.make_automaton()
        return auto


__all__ = [
    "GERMAN_NAMES",
    "GEORGIAN_NAMES",
    "PHIScrubber",
    "REDACTION",
    "SAFE_FIELD_NAMES",
    "ScrubberFailure",
    "phi_scrubber_failed_total",
]
